gluonnlp/model/sampled_block.py

```diff
@@ -0,0 +1,326 @@
+        Number of possible classes.
+    num_sampled: int
+        Number of classes randomly sampled for each batch.
+    in_unit: int
+        Dimensionality of the input space.
+    remove_accidental_hits: bool
+        Whether to remove "accidental hits" when a sampled candidate is equal to
+        one of the true classes.
+    """
+    def __init__(self, num_classes, num_sampled, in_unit, remove_accidental_hits,
+                 prefix=None, params=None):
+        super(_SampledLogits, self).__init__(prefix=prefix, params=params)
+        self._num_classes = num_classes
+        self._num_sampled = num_sampled
+        self._in_unit = in_unit
+        self._remove_accidental_hits = remove_accidental_hits
+
+    def hybrid_forward(self, F, x, sampled_candidates, expected_count_sampled,
+                       expected_count_true, label, w_all, b_all):
+        """Forward computation."""
+        # (num_sampled, in_unit)
+        w_sampled = w_all.slice(begin=(0, 0), end=(self._num_sampled, None))
+        w_true = w_all.slice(begin=(self._num_sampled, 0), end=(None, None))
+        b_sampled = b_all.slice(begin=(0,), end=(self._num_sampled,))
+        b_true = b_all.slice(begin=(self._num_sampled,), end=(None,))
+        # true logits
+        # (batch_size, 1)
+        x = x.reshape((-1, self._in_unit))
+        logits_true = (w_true * x).sum(axis=1) + b_true
+        # samples logits
+        # (batch_size, num_sampled)
+        b_sampled = F.reshape(b_sampled, (-1,))
+        logits_sampled = F.FullyConnected(x, weight=w_sampled, bias=b_sampled,
+                                          num_hidden=self._num_sampled)
+
+        # remove accidental hits
+        if self._remove_accidental_hits:
+            label_vec = F.reshape(label, (-1, 1))
+            sample_vec = F.reshape(sampled_candidates, (1, -1))
+            mask = F.broadcast_equal(label_vec, sample_vec) * -1e37
+            logits_sampled = logits_sampled + mask
+
+        # subtract log(q)
+        expected_count_sampled = F.reshape(expected_count_sampled,
+                                           shape=(1, self._num_sampled))
+        expected_count_true = expected_count_true.reshape((-1,))
+        logits_true = logits_true - F.log(expected_count_true)
+        logits_true = logits_true.reshape((-1, 1))
+        logits_sampled = F.broadcast_sub(logits_sampled, F.log(expected_count_sampled))
+
+        # logits and new_labels
+        # (batch_size, 1+num_sampled)
+        logits = F.concat(logits_true, logits_sampled, dim=1)
+        return logits, F.zeros_like(label)
+
+    def __repr__(self):
+        s = '{name}({mapping})'
+        mapping = '{0} -> {1}, with {2} samples'.format(self._in_unit, self._num_classes,
+                                                        self._num_sampled)
+        return s.format(name=self.__class__.__name__,
+                        mapping=mapping,
+                        **self.__dict__)
+
+class SampledLogits(HybridBlock):
```

Comment:
It would be better to rename this as ISNCELogitsNet, as the logit computation required negative samples which are specifically designed for IS or NCE, and are not approprite for other sampling methods. Besides, all the sampling method is useful in training, move the logits computation to model.train.